Add auto-labelling to pull requests

nilason opened the pull request:
This adds auto-labelling to pull requests (related to https://github.com/OSGeo/grass/issues/526).

It works with GitHub actions and add labels based on the PR's modified/added/deleted files location in repository.

This is a proposal, so if you find this approach appealing at all, I'm happy to receive suggestions for modifications.

Present proposal requires addition of some new labels.

The rules are stated in labeller.yml and as an example any modification in lib or include directory would add a `lib` label to the PR.

landam commented on .github/labeler.yml:
```diff
@@ -0,0 +1,38 @@
+lib:
```
This label doesn't exist in GitHub yet, see also related discussion about labels at, https://trac.osgeo.org/grass/wiki/GitMigration#Migrationoftracissues

neteler replied:
`lib` or `libraries`? I'd prefer the latter for being less cryptic.

nilason replied:
I chose `lib` as it's shorter, but `libraries` is good for me too.

landam commented on .github/labeler.yml:
```diff
@@ -0,0 +1,38 @@
+  - lib/**/*
+  - include/**/*
+
+core:
```
This label doesn't exist in GitHub yet, see also related discussion about labels at, https://trac.osgeo.org/grass/wiki/GitMigration#Migrationoftracissues

neteler replied:
It's `core` more like `modules`? To distinguish it from `lib(s)` or `libraries`?

nilason replied:
I was thinking of naming this modules, it might be a better choice.

landam commented on .github/labeler.yml:
```diff
@@ -0,0 +1,38 @@
+  - rpm/**/*
+  - singularity/**/*
+
+docker:
```
This label doesn't exist in GitHub yet, see also related discussion about labels at, https://trac.osgeo.org/grass/wiki/GitMigration#Migrationoftracissues

neteler replied:
`docker`is meanwhile there: https://github.com/OSGeo/grass/labels

landam commented on .github/labeler.yml:
```diff
@@ -0,0 +1,38 @@
+docker:
+  - docker/**/*
+
+docs:
```
We have already a label called `manual`. Either use `manual` or renamed existing label to `docs`.

wenzeslaus replied:
We were prefixing some commits with `docs:` and `doc:` in the past. `docs` is often used in Git repos for the documentation directory. `docs` might be broader than `manual`, so perhaps more useful as it can be applied to other things than just `doc` dir.

neteler replied:
Shall we rename the current label https://github.com/OSGeo/grass/labels/manual to docs? (fine for me)

landam commented on .github/labeler.yml:
```diff
@@ -0,0 +1,38 @@
+docs:
+  - doc/**/*
+
+locale:
```
This label doesn't exist in GitHub yet, see also related discussion about labels at, https://trac.osgeo.org/grass/wiki/GitMigration#Migrationoftracissues

neteler replied:
Is `locale` really clear to everybody or better a different word?

nilason replied:
not absolutely necessary to have this label at all...

wenzeslaus commented:
I think this would be useful to have or at least try out in action even if it is limited e.g. to the currently existing labels (if we want to keep the number of labels low).

neteler commented:
> I think this would be useful to have or at least try out in action even if it is limited e.g. to the currently existing labels (if we want to keep the number of labels low).

We may try, otherwise we'll never find out...

neteler commented:
@nilason does this PR need to be updated since GH Actions are now in place?

nilason commented:
> @nilason does this PR need to be updated since GH Actions are now in place?

No, this is a separate feature, should work as is.

neteler reviewed (dismissed):
Some comments added inline...

nilason commented:
Rebased and updated to better reflect current list of labels.

The labels `libraries` and `module` (or alternative label)  need to be created for this to work.

I commented out `docs` and `manual`, for the time being.

neteler commented:
PR rebased, but it needs more permissions:

```
Run actions/labeler@v4
...
Warning: The action requires write permission to add labels to pull requests. For more information please refer to the action documentation: https://github.com/actions/labeler#permissions
Error: Resource not accessible by integration
```

https://github.com/actions/labeler#permissions

nilason commented:
> PR rebased, but it needs more permissions:
> 
> ```
> Run actions/labeler@v4
> ...
> Warning: The action requires write permission to add labels to pull requests. For more information please refer to the action documentation: https://github.com/actions/labeler#permissions
> Error: Resource not accessible by integration
> ```
> 
> https://github.com/actions/labeler#permissions

Thanks!

Updated, works now on a private test repo of mine. (Looks like it have to be merged before kicking in).

wenzeslaus commented on .github/labeler.yml:
```diff
@@ -0,0 +1,75 @@
+---
+name: Labeler
+
+libraries:
```
We don't have this label, no? We do have C, C++ and Python which are used below.

nilason replied:
I believe my initial intention with `libraries` and `module`, was to be able to highlight evtl. changes to core vs. modules. Changes to libraries, being base to all, perhaps need particular attention.

wenzeslaus replied:
I think you need to create the labels first. I think they might be useful, so why not, especially if we have automation to help us label.

wenzeslaus commented on .github/labeler.yml:
```diff
@@ -0,0 +1,75 @@
+  - include/grass/**/*
+  - python/**/*
+
+module:
```
I can't find this label either. Maybe worth adding it. (?)

We have vector, raster, temporal, and database.

nilason replied:
Just added support to generate vector, raster, temporal, and database labels too.

wenzeslaus commented on .github/labeler.yml:
```diff
@@ -0,0 +1,75 @@
+docker:
+  - docker/**/*
+
+# docs:
+#   - doc/**/*
+# manual:
+#   - man/**/*
```
We have manual and HTML.

nilason replied:
Perhaps we can use the (existing) label `manual` for documentation in a broad sense like:
 
```
manual:
  - man/**/*
  - doc/**/*
```

wenzeslaus replied:
Or rename manual to documentation or docs. That would make sense to me.

nilason replied:
`docs` is short and concise! See also https://github.com/OSGeo/grass/pull/531#discussion_r422619437 and below...

nilason replied:
updated to use `docs`

wenzeslaus replied:
Sounds good. I agree with my 2020 self.

wenzeslaus commented on .github/labeler.yml:
```diff
@@ -0,0 +1,75 @@
+
+Linux:
+  - rpm/**/*
+  - singularity/**/*
```
Hm, docker or Linux label. I don't know.

neteler replied:
rpm could probably be skipped as it has never (?) received any reports in the past 10 years.

wenzeslaus commented on .github/labeler.yml:
```diff
@@ -0,0 +1,75 @@
+translation:
+  - locale/**/*
+
+CI:
+  - ".github/**/*"
```
Add `.travis`.

wenzeslaus commented on .github/labeler.yml:
```diff
@@ -0,0 +1,75 @@
+  - temporal/**/*
+  - vector/**/*
+
+# helperscripts:
+#  - utils/*
+#  - utils/**/*
```
Remove commented-out code, but you can include comment on what is included and what is not.

wenzeslaus commented:
> Updated, works now on a private test repo of mine.

I don't see any permissions in the YAML file, but I see them in the example in doc linked by @neteler. Any insights? What I'm asking is basically, do you expect the workflow to successfully run after merging it with the current configuration in terms of permissions?

> (Looks like it have to be merged before kicking in).

That's definitively the case.

echoix commented on .github/workflows/label.yml:
```diff
@@ -0,0 +1,19 @@
+  labeler:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/labeler@v4
```
```suggestion
      - uses: actions/labeler@v5
```
Latest action is now v5

echoix commented on .github/workflows/label.yml:
```diff
@@ -0,0 +1,19 @@
+# https://github.com/actions/labeler/blob/master/README.md
+
+name: "Pull Request Labeler"
+on: [pull_request_target]
```
```suggestion
on:
  - pull_request_target
```

echoix commented on .github/workflows/label.yml:
```diff
@@ -0,0 +1,19 @@
+on: [pull_request_target]
+
+jobs:
+  labeler:
```
```suggestion
  labeler:
    permissions:
      contents: read
      pull-requests: write
```

As the example of the action's docs.

echoix commented on .github/workflows/label.yml:
```diff
@@ -0,0 +1,19 @@
+
+name: "Pull Request Labeler"
+on: [pull_request_target]
+
```
```suggestion
permissions: {}

```

GitHub action permissions has evolved since, and the newest best practice is to specify at the workflow level the lowest permissions needed, and at the job level the minimum permissions needed. These permissions declare what the `GITHUB_TOKEN` is granted to do.


The advantage of setting the least at the workflow level is that further changes to the workflow by adding a job doesn't unintentionally give more rights that was previously planned out, but overlooked afterwards.


Declaring the permissions are quite easier to understand the config when looking at the workflow code, instead of a web interface that changes over time.

echoix commented on .github/workflows/label.yml:
```diff
@@ -0,0 +1,19 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/labeler@v4
+        with:
+          repo-token: "${{ secrets.GITHUB_TOKEN }}"
```
```suggestion
```
Not needed, it implicitly uses the GitHub token

echoix replied:
```suggestion
        with:
          sync-labels: true
```

Optionally, I think we could use the sync-labels option, and if it bothers us too much we (probably me, the first impacted), could remove it afterwards. 
https://github.com/actions/labeler#inputs

echoix reviewed (changes requested):
I left some new suggestions to have this PR up to date with changes with the action used. I would really like this PR to come through soon, it would be quite useful for tidying up the backlog (that I started doing)

echoix commented:
@neteler Just so you know, in this section it mentions that the action might fail one day at the time of a major version update of the action that would introduce breaking changes in the config, so that we might need to force it through while using the `pull_request_target` event type. That event type is really the most appropriate here however. They explain the workaround if ever needed.
https://github.com/actions/labeler#updating-major-version-of-the-labeler

echoix commented:
And for the different labels, and the old discussions about various small details, I think that it's overall quite adequate, docs is a good idea, since having details on html, css, JavaScript might be overkill.

However, if is possible to make a distinction between docs, the content, and docs, the system/infrastructure around it, that might be a better solution than having things like "manual".

Docs-only changes should be able to be merged more quickly without the same amount of programming-level code review. My stance is that it is already both hard to have contributors in open source and hard to have good docs in open source, so let's not leave out the contributions we could have. Thus, when a PR would have docs (the content) only changes, if it's clearer and no mistakes, it could be merged expeditiously. Docs, as the infrastructure, could be like any other topic, and I think there is work on changing from manually creating HTML to markdown, so these extra categories will change in the mid-future.


In conclusion, this PR, as a tool for us, should be fine tuned after usage, and it's our (changing) usage that dictates its requirements. We can't be that wrong with this. Our irritants will show us what to improve.

echoix commented on .github/labeler.yml:
I think that there might be missing a RFC filter, as they are being committed in the repo now, and should concern different members.

wenzeslaus replied:
I agree. The label is https://github.com/OSGeo/grass/labels/RFC

echoix commented:
Do I have the blessing to apply the suggestions I think are needed to unblock this PR?

nilason commented:
> Do I have the blessing to apply the suggestions I think are needed to unblock this PR?

I'm sitting with it at this moment.

nilason commented:
Thanks @echoix for the suggestions, added them with new commit.

nilason commented:
May I suggest to merge this ASAP, this workflow may be fine-tuned or even dropped is not good enough.

echoix commented:
I was about to finish approving it and go for merging. Merging main into the PR branch wasn't explicitly needed. We have enabled the option to always show it, so it was a suggestion rather than a warning, and it helps to retrigger CI for older PRs, without having to close/open the PR (which would run the same outdated CI), and rather pull in some new code that might include CI changes.

We'll see in 1h30+ if we can merge it then..
